src/streamlink/plugins/vimeo.py

```diff
@@ -8,70 +8,175 @@
 import logging
 import re
 from html import unescape as html_unescape
+from time import sleep, time
 from urllib.parse import urlparse
 
-from streamlink.plugin import Plugin, PluginArgument, PluginArguments, pluginmatcher
+from streamlink.plugin import Plugin, PluginArgument, PluginArguments, PluginError, pluginmatcher
 from streamlink.plugin.api import validate
 from streamlink.stream.dash import DASHStream
 from streamlink.stream.ffmpegmux import MuxedStream
-from streamlink.stream.hls import HLSStream
+from streamlink.stream.hls import HLSStream, MuxedHLSStream
 from streamlink.stream.http import HTTPStream
 
 log = logging.getLogger(__name__)
 
 
-@pluginmatcher(re.compile(
-    r"https?://(player\.vimeo\.com/video/\d+|(www\.)?vimeo\.com/.+)"
-))
-class Vimeo(Plugin):
-    _config_url_re = re.compile(r'(?:"config_url"|\bdata-config-url)\s*[:=]\s*(".+?")')
-    _config_re = re.compile(r"var\s+config\s*=\s*({.+?})\s*;")
-    _config_url_schema = validate.Schema(
-        validate.transform(_config_url_re.search),
-        validate.any(
-            None,
-            validate.Schema(
-                validate.get(1),
-                validate.parse_json(),
-                validate.transform(html_unescape),
-                validate.url(),
-            ),
-        ),
-    )
+class VimeoMuxedHLSStream(MuxedHLSStream):
+    def __init__(self, session, video, audio, **args):
+        super().__init__(session, video, audio, hlsstream=VimeoHLSStream, **args)
+
+
+class VimeoHLSStream(HLSStream):
+    __muxed__ = VimeoMuxedHLSStream
+
+    _expiry_time = 30
+    _expiry_re1 = re.compile(r"^exp=(\d+)~")
+    _expiry_re2 = re.compile(r"^(\d+)-")
+
+    def __init__(self, session_, url, **args):
+        self.session = session_
+        self._url = url
+        self._parsed_url = urlparse(self._url)
+        self._path_parts = self._parsed_url.path.split("/")
+        self._expiry = self.get_expiry(self._path_parts[1])
+        super().__init__(self.session, self._url, **args)
+
+    def _update_auth(self):
+        self.session.api._updating = True
+
+        data = Vimeo._get_data(self)
+        if not data:
+            raise ValueError("no video data found")
+        videos = data["request"]["files"]
+        if "hls" not in videos:
+            raise ValueError("hls key not found in video data")
+
+        new_auth = None
+        exit_loop = False
+        for _, video_data in videos["hls"]["cdns"].items():
+            url = video_data.get("url")
+            for _, stream in VimeoHLSStream.parse_variant_playlist(self.session, url).items():
+                parsed_url = urlparse(stream.to_manifest_url())
```

Comment:
You're parsing the variant playlist in order to get the URL of the variant playlist?

Btw, `dict.values()` only yields values, in contrast to `dict.items()` which yields tuples of keys and values.

Reply:
I should change this.  The two URLs are different because the URL passed to `parse_variant_playlist()` goes through a 302 redirect, but it doesn't seem efficient to process that in this way when a simple GET request can be used and the final URL can be obtained from that.